### ISSUE TITLE
My improved original model.

### ISSUE THREAD

**Issue by iperov** [deepfakes/faceswap-playground#120]
Before:

![out00168](https://user-images.githubusercontent.com/8076202/38174868-19cce71e-35e5-11e8-8be2-79020b7b14a8.png)
![2018-03-27_10-28-15](https://user-images.githubusercontent.com/8076202/38174869-1c254bf0-35e5-11e8-8419-3b30772a2294.jpg)

![2018-03-27_10-29-18](https://user-images.githubusercontent.com/8076202/38175188-810479ac-35e9-11e8-8605-53d72e6cd7c9.png)

After:

![out00168](https://user-images.githubusercontent.com/8076202/38174878-28b8b94c-35e5-11e8-817c-97348a2d25b6.png)
^ also this is 2GB low mem model, where reduced bottleneck dim and code dim.

What I changed: 
embedded local landmarks to png files
feeding network evenly with 18k sorted files by yaw
added non convex hull face mask to model ( 4th dimension input and output )
Fixed converter, there are two bugs in it. Now blur and erosion working properly, also histogram matches only masked face, so background color doesnt affect face color matching.

of course it works only in my DeepFaceSwapper global refactoring fork.

![nvidia share_2018-04-01_19-47-30](https://user-images.githubusercontent.com/8076202/38174941-e701a652-35e5-11e8-8221-523491e379c0.png)
(also there are many improvements versus original faceswap, such as multigpu extractor, automatic gpu and batch size manager in trainer, and other)
I did this in 5 days, while bryanlyon can only criticize my work and I was wasting time making prs
So no more criticism! Because I will not PR this. HAHAHAHAHHA

**Comment by iperov** [comment]
and again tons of words

**Comment by Kirin-kun** [comment]
Well, it looks cool. But coming here to taunt is a bit immature. Nobody stops you from making your own fork. Shaoanlu has his own work, dfaker has is own work, deepfakes has its own work. You have your own work. It's up to you to share it on github and lord over it. Or not.

Just curious, your masking uses face segmentation to exclude occlusions from conversion? If part of the face is missing.

**Comment by iperov** [comment]
you are wrong, nothing missing

**Comment by Kirin-kun** [comment]
You didn't understand my question. But whatever.

**Comment by iperov** [comment]
You didnt understand how original model works. But whatever,

**Comment by iperov** [comment]
Kirin-kun - is second reason why I will not share my work.

**Comment by iperov** [comment]
There are too many trolls in this community.

**Comment by gessyoo** [comment]
If you've created a better model, would you be willing share the code with those of us who don't code very well, maybe as a plugin model, of course with full credit to yourself as the author?

**Comment by iperov** [comment]
implemented adaptive blur+erosion size to perfect erode/blur mask on various face sizes

![fsviewer_2018-04-02_23-55-54](https://user-images.githubusercontent.com/8076202/38213384-0745abba-36d2-11e8-9e99-7eb8fa146067.png)
![fsviewer_2018-04-02_23-57-02](https://user-images.githubusercontent.com/8076202/38213386-094828e8-36d2-11e8-9e9b-54d4a7415fc6.png)

**Comment by iperov** [comment]
implemented loss history
![python_2018-04-02_23-59-41](https://user-images.githubusercontent.com/8076202/38213468-52931418-36d2-11e8-99fe-4694a6ecda2f.png)

**Comment by Kirin-kun** [comment]
At this point, it's a new project and I think you're better off doing your own thing, because you're not suited for collaborative work with the people here.

Good luck in your endeavor!

**Comment by iperov** [comment]
im going to port DF to my architecture. It has better training data generator algorithm - warping src face to dst face by landmarks and randomizing them.

**Comment by iperov** [comment]
lol why close in playground ?

**Comment by deepfakesclub** [comment]
I haven't commented before about this because I'm a mere end user, but come on. There are unhelpful and inactive old issues currently still open, but this one is closed?

I don't agree with his antisocial attitude, and I don't understand all of the technical details because I'm not a programmer. However, for the technical and non-technical parts that I do understand, I have mostly agreed with @iperov. People have been constantly jumping all over his PRs since day 1 - are you surprised that he is annoyed after all the work he put in? And these days inferior PRs are committed with barely a discussion? He has every right to be upset, although no excuse for how he dealt with it. Voices from people who contribute more should be given more leeway and weight - I mean before the insults, not after. I've specifically refrained from commenting on other PRs for this reason.

Let's be clear. There are inept programmers contributing bad code. There are inept managers misdirecting the whole affair. There are inept users wasting people's time. And there are socially inept people tangled in the mix. But who cares. We're not saving lives here. We're pursuing a leisure hobby. Plus, all of you are a pretty cool group. I'll be more explicit. I think @bryanlyon has made illogical statements, but I don't care, because I still think @bryanlyon is a good guy to have around.

My only major difference of opinion with @iperov is that I think all sides are doing work that can provide value and education. Yes, the repo is growing like a cancer, but that just means more ideas and code snippets to learn from. I think it's great to cultivate all the programmers that want to contribute, and that includes @iperov. He is taking ideas from here (please note, licensed by GNU GPL3) to write his own code. People can take ideas from him as well. @Clorr had the embedded png idea and @iperov implemented it. Good things happen by sharing ideas, not just code.

And no, there is zero need to kiss up to him. The advances these days are incremental, and if he quits, it's not a major blow to the world or anything. None of what has been proposed or done is a game-changing advance, and definitely not going to bring about world peace. I'm pretty sure he called me dumb in the past as well. I'm just calling it as I see it.

It seems like @iperov is willing to tone down his antics and go back to focusing on the technical side. I get that people are emotional humans, but @iperov deserves some credit for what he has contributed. I make no excuses for him, but I have also seen him genuinely try to help people before, too. So I'll at least chime in my two cents on his sort-of behalf. I actually think splitting up into community and solo forks is a good idea and will make for a stronger ecosystem, especially if there is crosstalk and communication.

**Comment by Kirin-kun** [comment]
Like @iperov would say: too many words.

Everyone is free. He intends to develop his own fork and I approve. If it's better, people interested will migrate toward his project and eventually contribute. It will help everyone.

**Comment by deepfakesclub** [comment]
Then, he has a right to discuss his related work here.

**Comment by Kirin-kun** [comment]
Discussing his improvements here is taken as gloating by some people. I think he just wants validation/acknowledgement of his work. But at this point it's a clash of egos and this is too far removed from programming to be interesting.

**Comment by iperov** [comment]
I dont need acknowledgement of my work from bunch of noobs like you.
My goal make better program for end users and better platform for programmers.

**Comment by deepfakesclub** [comment]
@Kirin-kun Understood. Similar boat on the time issue for most of us...

@iperov You don't owe anyone a damn thing. But, please?

Okay, I'm done.  Yes, this is likely a waste of time. But I beg to differ on your previous point. Programmers will always be more important than programming. Worth one, and only one, try.

**Comment by iperov** [comment]
there are only 3 three people annoying me in this repo: bryanlon, kirinkun and apollo122

**Comment by iperov** [comment]
deepfakesclub thx for promoting deepfakes

**Comment by Apollo122** [comment]
@iperov now that you are no longer doing PRs, and using your own code base what's your reason posting stuff here?

**Comment by bryanlyon** [comment]
My reason for closing this was that this was just gloating and attacking.  He refused to share anything he's done and there was no real discussion.  I'm fine with it reopening since it's now encouraging actual discussion.

**Comment by iperov** [comment]
I will share it anyway, because code anyway will be exist in windows release, so why hide it.

**Comment by iperov** [comment]
currently I testing MTCNN vs DLIBCNN.
In first look MTCNN better, but it finds too many false faces, which not problem with my sorttool.
